src/utils/Matcher.ts

```diff
@@ -0,0 +1,149 @@
+import UserRepo from '../repos/UserRepo';
+import User from '../entities/User';
+
+// MATCHER CONSTANTS
+
+const AVAILABILITY_SCORE_MAX = 20;
+const MAJOR_SCORE_MAX = 20;
+const INTEREST_SCORE_MAX = 20;
+const GROUP_SCORE_MAX = 20;
+const TALKING_POINTS_SCORE_MAX = 20;
+
+// PREVIOUS WEEK FUNCTIONS
+
+async function finalizeMatchStatuses() {
+  const previousWeeksMatches = await MatchRepo.getWeeklyMatches();
+  previousWeeksMatches.forEach((match) => {
+    const status =
+      match.status === Constants.MATCH_ACTIVE ? Constants.MATCH_INACTIVE : Constants.MATCH_CANCELED;
+    MatchRepo.updateMatch(match, { status });
+  });
+}
+
+// SIMILARITY FUNCTIONS
+
+const calculateSimilarity = (chooserItems: any[], choiceItems: any[]) => {
+  let similarity = 0;
+  chooserItems.forEach((item) => {
+    similarity += item in choiceItems ? 1 : 0;
+  });
+  return chooserItems.length / similarity;
+};
+
+function generalListScorer(chooser: User, choice: User, property: string, max: number) {
+  let score = max;
+  let similarity;
+  if (Constants.FINDING_MY_PEOPLE in chooser.goals) {
+    similarity = calculateSimilarity(chooser[property], choice[property]);
+    score -= (max / 2) * (1 - similarity);
+  }
+  if (Constants.MEETING_SOMEONE_DIFFERENT in chooser.goals) {
+    const similarityValue = similarity || calculateSimilarity(chooser[property], choice[property]);
+    score -= (max / 2) * similarityValue;
+  }
+  return score;
+}
+
+function availabilityScorer(chooser: User, choice: User) {
+  if (chooser.availabilities.length === 0) return 0;
+  let similarity = 0;
+  chooser.availabilities.forEach((chooserAvailability) => {
+    let currentSimilarity = 1;
+    choice.availabilities.forEach((choiceAvailability) => {
+      if (chooserAvailability.day === choiceAvailability.day) {
+        currentSimilarity = calculateSimilarity(
+          chooserAvailability.times,
+          choiceAvailability.times,
+        );
+      }
+    });
+    similarity += currentSimilarity;
+  });
+  return (AVAILABILITY_SCORE_MAX * similarity) / chooser.availabilities.length;
+}
+
+function majorScorer(chooser: User, choice: User) {
+  let score = MAJOR_SCORE_MAX;
+  if (Constants.FINDING_MY_PEOPLE in chooser.goals) {
+    score -= chooser.major !== choice.major ? MAJOR_SCORE_MAX / 2 : 0;
+  }
+  if (Constants.MEETING_SOMEONE_DIFFERENT in chooser.goals) {
+    score -= chooser.major === choice.major ? MAJOR_SCORE_MAX / 2 : 0;
+  }
+  return score;
+}
+
+function interestScorer(chooser: User, choice: User) {
+  return generalListScorer(chooser, choice, 'interests', INTEREST_SCORE_MAX);
+}
+
+function groupScorer(chooser: User, choice: User) {
+  return generalListScorer(chooser, choice, 'groups', GROUP_SCORE_MAX);
+}
+
+function talkingPointsScorer(chooser: User, choice: User) {
+  return generalListScorer(chooser, choice, 'talkingPoints', TALKING_POINTS_SCORE_MAX);
+}
+
+// MATCHING FUNCTIONS
+
+async function createUserPreferences(): Promise<Record<string, User[]>> {
```

Comment:
Maybe considering adding a specification to this function to make it easier to digest especially is uses the scorers defined earlier?